<?xml version="1.0"?>
<launch>
  <arg name="gui" default="false"/>
  <arg name="max_velocity_scale_factor" default="1.0" doc="Scale factor multplied by maximum joint velocity value"/>
  <arg name="server_port" default ="50051"/>
  <arg name="action_cycle_rate" default="25" doc="Rate at which new commands are published to the robot controller"/>
  <arg name="reference_frame" default="base" doc="Reference frame with respect to which target and ee pose are given"/>
  <arg name="target_mode" default="fixed" doc="fixed, moving"/>

  <!-- Scene Objects -->
  <arg name="objects_controller" default="false"  doc="Start objects controller node"/>
  <arg name="n_objects" default="0.0"  doc="Number of objects in the scene"/>
  <arg name="object_trajectory_file_name" default="no_file" doc="Object trajectory file name"/>
  <arg name="object_0_model_name" default="" doc="Gazebo model name of the target object"/>
  <arg name="object_0_frame" default="" doc="TF Frame name of target"/>
  <arg name="object_1_model_name" default="" doc="Gazebo model name of object 1"/>
  <arg name="object_1_frame" default="" doc="TF Frame name of object 1"/>

  <!-- RViz Configuration -->
  <arg name="rviz_config_path" default="$(find ur_robot_server)/rviz"/>
  <arg name="rviz_config_file" default="ur_rob_rl.rviz" doc="RViz configuration file"/>
  <arg name="rviz_config" value="$(arg rviz_config_path)/$(arg rviz_config_file)"/>


  <!-- Launch Command Handler  -->
  <node name="joint_trajectory_command_handler" pkg="ur_robot_server" type="joint_trajectory_command_handler.py" respawn="false" output="screen">
    <param name="real_robot" value="true"/>
    <param name="action_cycle_rate" value="$(arg action_cycle_rate)"></param>
  </node>

  <!-- Launch RViz -->
  <node if="$(arg gui)" name="$(anon rviz)" pkg="rviz" type="rviz" respawn="false"
  args="-d $(arg rviz_config)" output="screen">
  </node>

  <!--Launch gRPC server -->
  <node name="robot_server" pkg="ur_robot_server" type="robot_server.py" respawn="false" output="screen">
    <param name="server_port" value="$(arg server_port)"></param>
    <param name="real_robot" value="true"></param>
    <param name="ur_model" value="ur5"></param>
    <param name="max_velocity_scale_factor" value="$(arg max_velocity_scale_factor)"></param>
    <param name="action_cycle_rate" value="$(arg action_cycle_rate)"></param>
    <param name="reference_frame" value="$(arg reference_frame)"></param>
    <param name="target_mode" value="$(arg target_mode)"></param>
  </node>

  <!-- Parameters -->
  <param name="real_robot" value="false"></param>
  <param name="reference_frame" value="$(arg reference_frame)"></param>
  <!-- Scene Objects Parameters -->
  <param name="objects_controller" value="$(arg objects_controller)"></param>
  <param name="object_trajectory_file_name" value="$(arg object_trajectory_file_name)"></param>
  <param name="n_objects" value="$(arg n_objects)"></param>
  <param name="object_0_model_name" value="$(arg object_0_model_name)"></param>
  <param name="object_0_frame" value="$(arg object_0_frame)"></param>
  <param name="object_1_model_name" value="$(arg object_1_model_name)"></param>
  <param name="object_1_frame" value="$(arg object_1_frame)"></param>

  <!--Launch objects controller -->
  <node if="$(arg objects_controller)" name="objects_controller" pkg="ur_robot_server" type="objects_controller.py" respawn="false" output="screen" ></node>

</launch>
